### ISSUE TITLE
Debugger for scala in vscode

### ISSUE THREAD

**Issue by virtualdvid** [scalameta/metals#1058]
I have been searching a lot how to debug scala in vscode without success. Wondering if there is a way or if you guys are working on that at the moment?.

**Comment by olafurpg** [comment]
Since https://github.com/scalacenter/bloop/pull/1087 was merged into Bloop, it's possible to breakpoint debug Scala/Java sources that are defined in the workspace. After #1010, it will be possible to debug 3rd party libraries as well. Once that's merged then this issue can be closed.

**Comment by ctataryn** [comment]
I'm curious, the Bloop documentation states:

_If you want to debug your code with Bloop via an editor such as VS Code, vim, Sublime or emacs you can use Metals._ [1]

Are they just jumping the gun on their decree?

[1] - https://scalacenter.github.io/bloop/docs/debugging-reference#debugging-via-an-editor

**Comment by tgodzik** [comment]
@ctataryn It's not yet possible, from what I see we still need to finish a couple of things.

**Comment by gabro** [comment]
v0.8.0 is out, supporting this feature! Check out the release notes https://scalameta.org/metals/blog/2020/01/10/cobalt.html#debugging

**Comment by virtualdvid** [comment]
Hi, Thank you for these release! I'm getting an issue.

I created a `hello.scala` script to test it. In the `cmd`:

`C:\Users\myuser\project\scala\my_project>scala hello.scala
Hello World!`

But when I use the debugger option in vscode nothing happens:

![scala metals](https://user-images.githubusercontent.com/33104726/73084989-d6e19500-3e82-11ea-9544-0581e8da2ff4.gif)

This is happening in my personal and job laptop.

**Comment by ctataryn** [comment]
@virtualdvid when I debug, what happens is above the `Hello` object there’d be a “Run | Debug” insight that appears.

Click on the “M” icon on the left toolbar, then run the “Doctor” to see if there are any problems. I know for me it wasn’t until I upgraded my project’s Scala version that it all started working.

**Comment by virtualdvid** [comment]
Hi @ctataryn, Thanks for your answer. I got several bugs solved. I did it from scratch since uninstalling vscode. I'm getting now all test in `doctor` passing:

![Code_yrPDVowrUh](https://user-images.githubusercontent.com/33104726/73106778-16bf7100-3eb1-11ea-9017-996253074d56.png)

Same result as before, the debugger is not running.

**Comment by ctataryn** [comment]
@virtualdvid Hmmm, I'm not entirely sure then. What you could do is exit VSCode, remove the `.metals` folder, then re-start VSCode and the extension should ask you about "importing" your project again.

As a side note: I just opened VSCode an my Main.scala file was auto-loaded in a tab.  I noticed that the `Run | Debug` options didn't appear until I switched to a different file, and then went back to Main.scala.

**Comment by virtualdvid** [comment]
@ctataryn I did what you said same result in both laptops. Can you share your `launch,json` file. I'm curious how that looks like. Thanks!

**Comment by ctataryn** [comment]
@virtualdvid actually the extension doesn't use `launch.json` and I don't recognize the "Debug and Run Start" pane you have there on the left. Perhaps it is another extension that's interfering?  Here's a shot of extensions that I have installed, I'd suggest disabling all but Metals and any other essential extensions you may need.

And also, double check to see that your Metals extension is at `v1.7.2`

![image](https://user-images.githubusercontent.com/435719/73117498-1d2f0680-3f0c-11ea-862e-cf8cb716e391.png)

**Comment by virtualdvid** [comment]
@ctataryn Thanks for your help. I don't know what happened. I turned on my job laptop today and it is working. In my personal laptop stills the issue. Weird!

**Comment by tgodzik** [comment]
`WARN no build target for: C:\Users\david\Documents\Documents\scala\my_proyect\src\main\hello.scala` usually means that the build was either not imported of the directory structure does not correspond to that exported from the build tool.

**Comment by olafurpg** [comment]
Thank you everybody for reporting back feedback! I encourage you to open a new issue to discuss bugs/problems you encounter. This issue here was mainly about introducing basic debugging functionality.

Debugging is a new feature with a large surface area so it's better to have focused issues on specific functionality. We are also aware that there are also some important features missing like run configurations (see https://github.com/scalameta/metals-feature-requests/issues/63)

**Comment by mrseanryan** [comment]
Am having a similar issue with debug not working - it seems to connect to the debug session, but never reaches the 1st breakpoint.  It just sits there:

Output pane:

```
INFO  time: compiled forcomp in 0.95s
INFO  Listening for transport dt_socket at address: 56516
INFO  tracing is enabled: C:\Users\str_i\AppData\Local\scalameta\metals\cache\dap-server.trace.json
INFO  tracing is enabled: C:\Users\str_i\AppData\Local\scalameta\metals\cache\dap-client.trace.json
INFO  Starting debug proxy for [Main]
INFO  Trying to attach to remote debuggee VM localhost:56516 .
INFO  Attaching to debuggee VM succeeded.
```

note: I run debug by clicking the 'debug' link above the main entry point:

![image](https://user-images.githubusercontent.com/16241449/75090653-4c658180-5565-11ea-8e79-7c7323c0242c.png)

note: if I click on `run` it does run OK.

In the log there is a similar warning:

```
WARN  no build target for: C:\sean\github\mrseanryan\scala-learning\projects\week6\tryout\src\main\scala\tryout\Main.scala
```

So maybe my project structure is incorrect 🤔 

Perhaps someone could point me to a small example project, where debugging works ok ?

btw thanks for making this extension!

**Comment by olafurpg** [comment]
Please open new tickets to start new discussions or ask questions.
